nx/lib/nx/serving.ex

```diff
@@ -283,6 +283,45 @@ defmodule Nx.Serving do
   In partitioned mode, `c:init/3` may receive multiple `defn_options`
   as the third argument and `c:handle_batch/3` may receive another partition
   besides 0.
+
+  ### Batch keys
+
+  Sometimes it may be necessary to execute different functions under the
+  same serving. For example, sequence transformers must pad the sequence
+  to a given length. However, if you are batching, the length must be
+  padded upfront. If the length is too small, you have to either discard
+  data or support only small inputs. If the length is too large, then you
+  decrease performance with the extra padding.
+
+  Batch keys provide a mechanism to accumulate different batches, based on
+  their key, which execute independently. As an example, we will do a
+  serving which performs different operations based on the batch key,
+  but it could also be used to perform the same operation for different
+  templates:
+
+      iex> args = [Nx.template({10}, :s64)]
+      iex> serving = Nx.Serving.new(fn
+      ...>   :double, opts -> Nx.Defn.compile(&Nx.multiply(&1, 2), args, opts)
+      ...>   :half, opts -> Nx.Defn.compile(&Nx.divide(&1, 2), args, opts)
+      ...> end)
+      iex> double_batch = Nx.Batch.concatenate([Nx.iota({10})]) |> Nx.Batch.key(:double)
+      iex> Nx.Serving.run(serving, double_batch)
+      #Nx.Tensor<
+        s64[10]
+        [0, 2, 4, 6, 8, 10, 12, 14, 16, 18]
+      >
+      iex> half_batch = Nx.Batch.concatenate([Nx.iota({10})]) |> Nx.Batch.key(:half)
+      iex> Nx.Serving.run(serving, half_batch)
+      #Nx.Tensor<
+        f32[10]
+        [0.0, 0.5, 1.0, 1.5, 2.0, 2.5, 3.0, 3.5, 4.0, 4.5]
+      >
```

Comment:
I think you should also include an example for same batch key but different input shape. The paragraphs above mention this, but the only examples are in the unit tests.

Reply:
It is tricky because the benefit of different shapes only happen in process mode and then the example becomes overly long (and you can't run as easily).